src/rpc/tikv/client.rs

```diff
@@ -556,7 +556,19 @@ impl KvClient {
         start_key
             .map(|k| req.set_start_key(k.into_inner()))
             .unwrap();
-        end_key.map(|k| req.set_end_key(k.into_inner())).unwrap();
+        end_key
+            .map(|k| {
+                // Scan in TiKV has a weird particularity to it.
+                // The **start** of a scan is inclusive, unless appended with an '\0', then it is exclusive.
+                // The **end** of a scan is exclusive, unless appended with an '\0', then it is inclusive.
```

Comment:
I would expect the scan range to be inclusive/exclusive and if TiKV does that, then the client should too, I think. Using the `\0`s is weird and we should work around it. Also Rust ranges are by default inclusive/exclusive.

Reply:
The Client handles ranges from all external code by taking `x..y` or `x..=y` or whatever else. The user doesn't need to think  about it. The way this works is by always building a set of tuples that create (inclusive) bounds and pass those around. This means all internal APIs just deals with inclusive range bounds instead of range type objects. (This is primarily because ranges are challenging to work with if you want to avoid copying all over the place. So we just copy once to take ownership and pass it around

Reply:
Hmm, so that seems like a mistake in the implementation of `KeyRange`, but we should probably land this as is and address that later (I have also refactored `KeyRange` in one of the commits I have queued up, so I think we can land this and that one and then look at `KeyRange`'s treatment of the end key).

Reply:
What seems like a mistake in `KeyRange`?

Reply:
Having the upper bound be inclusive